### PR TITLE
FOUR-9610: Duplicate Project Assets

### PR DESCRIPTION
**Pull request by estebangallego**
## Issue & Reproduction Steps
As a Solution Developer, I want the ability to duplicate an asset while retaining all of its characteristics, ensuring that I can make edits without affecting other projects.

## Solution
- Implement asset duplication within the Project

## How to Test
- Checkout to the FOUR-9610 branch in Core and Package Projects
- Create a few Projects 
- Add an Asset to the Project
- Navigate to the Project
- Click on the Duplicate within the Elipsis Menu 
- Fill up the inputs and click on Duplicate
----------------
- Please ensure that the Asset has been duplicated.
- If a Process has Screens and Scripts, those assets should be duplicated and added to the project.
- We can also Duplicate assets to multiple projects.
- Please ensure that the unit test passes as well.
-` phpunit vendor/processmaker/package-projects/tests/Models/AssetsTest.php --filter=testDuplicateAssetFromProject`

## Related Tickets & Packages
- Ticket [FOUR-9610](https://processmaker.atlassian.net/browse/FOUR-9610)
- Package Projects

## CI 
ci:package-projects:task/FOUR-9610
ci:next
ci:deploy

## Code Review Checklist
- [ ] I have pulled this code locally and tested it on my instance, along with any associated packages.
- [ ] This code adheres to [ProcessMaker Coding Guidelines](https://github.com/ProcessMaker/processmaker/wiki/Coding-Guidelines).
- [ ] This code includes a unit test or an E2E test that tests its functionality, or is covered by an existing test.
- [ ] This solution fixes the bug reported in the original ticket.
- [ ] This solution does not alter the expected output of a component in a way that would break existing Processes.
- [ ] This solution does not implement any breaking changes that would invalidate documentation or cause existing Processes to fail.
- [ ] This solution has been tested with enterprise packages that rely on its functionality and does not introduce bugs in those packages.
- [ ] This code does not duplicate functionality that already exists in the framework or in ProcessMaker.
- [ ] This ticket conforms to the PRD associated with this part of ProcessMaker.

[FOUR-9610]: https://processmaker.atlassian.net/browse/FOUR-9610?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: ProcessMaker/Http/Controllers/Api/ExportController.php [hunk 1]
```diff
@@ -19,6 +18,10 @@
 use ProcessMaker\Models\ProcessTemplates;
 use ProcessMaker\Models\Screen;
 use ProcessMaker\Models\Script;
+use ProcessMaker\Package\PackageDecisionEngine\ImportExport\DecisionTableExporter;
+use ProcessMaker\Package\PackageDecisionEngine\Models\DecisionTable;
+use ProcessMaker\Packages\Connectors\DataSources\ImportExport\DataSourceExporter;
+use ProcessMaker\Packages\Connectors\DataSources\Models\DataSource;
```
Comment (sanjacornelius) [inline ProcessMaker/Http/Controllers/Api/ExportController.php:24]: Hey @estebangallego before initializing these classes, could we implement a check to verify if the corresponding packages are installed or if the classes exist? This precautionary step would help avoid potential issues for instances that do not have these packages installed or for newly spun-up instances. 

So for example:

```
use ProcessMaker\PackageHelper;

const DATA_SOURCE_CLASS = 'ProcessMaker\Packages\Connectors\DataSources\Models\DataSource';
const DATA_SOURCE_EXPORTER_CLASS = 'ProcessMaker\Packages\Connectors\DataSources\ImportExport\DataSourceExporter';

const DECISION_TABLE_CLASS = 'ProcessMaker\Package\PackageDecisionEngine\Models\DecisionTable';
const DECISION_TABLE_EXPORTER_CLASS = 'ProcessMaker\Package\PackageDecisionEngine\ImportExport\DecisionTableExporter';

const IS_DATA_SOURCES_INSTALLED = PackageHelper::isPackageInstalled(self::DATA_SOURCE_CLASS);
const IS_DECISION_TABLE_INSTALLED = PackageHelper::isPackageInstalled(self::DECISION_TABLE_CLASS);

class ExportController extends Controller
{
    protected array $types = [
        'process' => [Process::class, ProcessExporter::class],
        'script' => [Script::class, ScriptExporter::class],
        'process_templates' => [ProcessTemplates::class, TemplateExporter::class],
        'decisiontable' => IS_DECISION_TABLE_INSTALLED ? [self::DECISION_TABLE_CLASS, self::DECISION_TABLE_EXPORTER_CLASS] : null,
        'datasource' => IS_DATA_SOURCES_INSTALLED ? [self::DATA_SOURCE_CLASS, self::DATA_SOURCE_EXPORTER_CLASS] : null,
    ];
}
```
Let me know if you have questions or run into problems with the above code

**Review by sanjacornelius — CHANGES_REQUESTED** [review]

**Review by sanjacornelius — APPROVED** [review]

**Comment by tokensPM** [comment]
QA server was successfully deployed https://ci-56cfc673bb.eng.processmaker.net

File: ProcessMaker/Http/Controllers/Api/ExportController.php [hunk 2]
```diff
@@ -19,16 +18,35 @@
...
         'process_templates' => [ProcessTemplates::class, TemplateExporter::class],
     ];
 
+    public function __construct()
+    {
+        if (PackageHelper::isPackageInstalled(self::DATA_SOURCE_CLASS)) {
+            $this->types['datasource'] = [self::DATA_SOURCE_CLASS, self::DATA_SOURCE_EXPORTER_CLASS];
+        }
+        if (PackageHelper::isPackageInstalled(self::DECISION_TABLE_CLASS)) {
+            $this->types['decisiontable'] = [self::DECISION_TABLE_CLASS, self::DECISION_TABLE_EXPORTER_CLASS];
```
Comment (eiresendez) [inline ProcessMaker/Http/Controllers/Api/ExportController.php:46]: This PR has already been approved, but I wonder if not following the snake_case convention could bring problems in the future? 

- `datasource` - `data_source` and `decisiontable` - `decision_table`.

Also, keep in mind that from the front-end it is sent as kebab-case:

`<add-to-project-modal id="add-to-project-modal" ref="add-to-project-modal"  assetType="data-source" :assetId="assetId" :assetName="assetName" />`


cc: @sanjacornelius @estebangallego

**Review by eiresendez — APPROVED** [review]

File: ProcessMaker/Http/Controllers/Api/ExportController.php (hunk 2 above)
Comment (estebangallego) [reply ProcessMaker/Http/Controllers/Api/ExportController.php]: Resolved! Thank you @eiresendez

**Comment by tokensPM** [comment]
QA server was successfully deployed https://ci-56cfc673bb.eng.processmaker.net

**Pull request merged**